File: benchmark_scripts/mergesort-mpi/jl/mergesort.jl
using MPI
MPI.Init()
comm = MPI.COMM_WORLD
rank = MPI.Comm_rank(comm)
size = MPI.Comm_size(comm)

# Insertion sort for small arrays (right exclusive) --> the leaves of merge sort
function insertion_sort!(arr, left, right)
    for i in left+1:right-1
        key = arr[i]
        j = i - 1
        while j ≥ left && arr[j] > key
            arr[j+1] = arr[j]
            j -= 1
        end
        arr[j+1] = key
    end
end

# Merge step a[left:mid] and a[mid:right] using tmp buffer
function merge!(arr, left, mid, right, tmp)
    i = left
    j = mid
    k = left

    while i < mid && j < right
        if arr[i] <= arr[j]
            tmp[k] = arr[i]
            i += 1
        else
            tmp[k] = arr[j]
            j += 1
        end
        k += 1
    end

    while i < mid
        tmp[k] = arr[i]
        i += 1
        k += 1
    end

    while j < right
        tmp[k] = arr[j]
        j += 1
        k += 1
    end

    # Copy back
    @inbounds for idx in left:right-1
        arr[idx] = tmp[idx]
    end
end

# Recursive merge sort implementation
function merge_sort!(arr, left, right, tmp, threshold)
    size = right - left
    if size <= 1
        return
    end

    if size <= threshold
        insertion_sort!(arr, left, right)
        return
    end

    mid = left + size ÷ 2

    merge_sort!(arr, left, mid, tmp, threshold)
    merge_sort!(arr, mid, right, tmp, threshold)
    merge!(arr, left, mid, right, tmp)
end

# Compute child rank in binary tree of ranks
function compute_child(rank, depth, max_depth)
    if depth >= max_depth
        return nothing
    end
    bit_pos = max_depth - depth - 1
    child = rank | (1 << bit_pos)

    if child >= size
        return nothing
    end
    if child == rank
        return nothing
    end
    return child
end

# Parallel merge sort implementation sending to ranks in recursion --> binary tree of ranks
# only the leaves do perform work (recursive merge sort, insertion sort at the bottom)
function merge_sort_parallel(arr, threshold, max_depth, depth)
    n = length(arr)

    child = compute_child(rank, depth, max_depth)
    if child === nothing || n <= threshold
        tmp = copy(arr)
        merge_sort!(arr, 1, n+1, tmp, threshold)
        return arr
    end

    mid = div(n, 2)
    left_part  = arr[1:mid]
    right_part = arr[mid+1:end]

    # send right part to child
    MPI.send(right_part, comm; dest=child, tag=depth)

    # recurse on left
    sorted_left = merge_sort_parallel(left_part, threshold, max_depth, depth + 1)

    # receive right sorted part from child (source/tag known)
    sorted_right = MPI.recv(comm; source=child, tag=depth)

    # merge
    merged = vcat(sorted_left, sorted_right)
    tmp = copy(merged)
    merge!(merged, 1, length(sorted_left)+1, length(merged)+1, tmp)

    return merged
end

# Main
function main()
    if length(ARGS) != 2
        println("Usage: julia mergesort.jl SIZE THRESHOLD")
        MPI.Finalize()
        exit()
    end

    N = parse(Int, ARGS[1])
    threshold = parse(Int, ARGS[2])
    max_depth = floor(Int, log2(size))    

    if rank == 0
        # Generate reverse sorted array
        start_gen = time_ns()
        arr = [N - i for i in 0:N-1]

        start_sort = time_ns()
        sorted_arr = merge_sort_parallel(arr, threshold, max_depth, 0)
        stop_time =time_ns()

        # Correctness check: H = Sum^{N-1}_{i=0} (i * arr[i}) <=> H = N(N-1)(N+1)/3
        H = 0
        for i in 1:length(arr)
            H += (i-1) * sorted_arr[i]     
        end
        c = H / (N * N * N)

        duration_ete = (stop_time - start_gen) * 1e-9
        duration_sort = (stop_time - start_sort) * 1e-9

        println("$(duration_ete),$(duration_sort),$(c)")

        # terminate workers
        for r in 1:size-1
            MPI.send(Int[], comm; dest=r, tag=999)
        end

    else
        while true
            # receive message from any parent, with status to get tag/source
            msg, status = MPI.recv(comm, MPI.Status;
                                   source=MPI.ANY_SOURCE, tag=MPI.ANY_TAG)
            parent = MPI.Get_source(status)
            tag    = MPI.Get_tag(status)

            if tag == 999
                break
            end

            depth = tag
            data = msg::Vector{Int}
            sorted_data = merge_sort_parallel(data, threshold, max_depth, depth + 1)
            MPI.send(sorted_data, comm; dest=parent, tag=depth)
        end
    end
end

main()
MPI.Finalize()

File: benchmark_scripts/mergesort-mpi/py/mergesort.py
import math
import sys
import time
from mpi4py import MPI

comm = MPI.COMM_WORLD
rank = comm.Get_rank()
size = comm.Get_size()

# Insertion sort for small arrays (right exclusive) --> the leaves of merge sort
def insertion_sort(arr, left, right):
    for i in range(left + 1, right):
        key = arr[i]
        j = i - 1
        while j >= left and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key

# Merge step a[left:mid] and a[mid:right] using tmp buffer
def merge(arr, left, mid, right, tmp):
    i, j, k = left, mid, left

    while i < mid and j < right:
        if arr[i] <= arr[j]:
            tmp[k] = arr[i]
            i += 1
        else:
            tmp[k] = arr[j]
            j += 1
        k += 1

    while i < mid:
        tmp[k] = arr[i]
        i += 1
        k += 1

    while j < right:
        tmp[k] = arr[j]
        j += 1
        k += 1

    # copy back
    for idx in range(left, right):
        arr[idx] = tmp[idx]

# Recursive merge sort implementation
def merge_sort(arr, left, right, tmp, threshold):
    arr_size = right - left
    if arr_size <= 1:
        return

    if arr_size <= threshold:
        insertion_sort(arr, left, right)
        return

    mid = left + arr_size // 2

    merge_sort(arr, left, mid, tmp, threshold)
    merge_sort(arr, mid, right, tmp, threshold)
    merge(arr, left, mid, right, tmp)

# Compute child rank based on current depth
def compute_child(rank, depth, max_depth):
    if depth >= max_depth:
        return None
    bit_pos = max_depth - depth - 1
    child = rank | (1 << bit_pos)
    if child >= size:
        return None
    if child == rank:
        return None
    return child

# Parallel merge sort implementation sending to ranks in recursion --> binary tree of ranks
# only the leaves do perform work (recursive merge sort, insertion sort at the bottom)
def merge_sort_parallel(arr, threshold, max_depth, depth):
    n = len(arr)
    
    # compute child rank
    child = compute_child(rank, depth, max_depth)
     # no child available (=leaf) or below threshold --> local merge sort
    if child is None or n <= threshold:
        tmp = arr.copy()
        merge_sort(arr, 0, n, tmp, threshold)
        return arr
    
    # split array
    mid = n // 2
    left_part = arr[:mid]
    right_part = arr[mid:]

    # send right part to child
    comm.send(right_part, dest=child, tag=depth)

    # sort left part locally
    sorted_left = merge_sort_parallel(left_part, threshold, max_depth, depth + 1)

    # receive sorted right part from child
    sorted_right = comm.recv(source=child, tag=depth)

    # merge sorted parts
    merged = sorted_left + sorted_right
    tmp = merged.copy()
    merge(merged, 0, len(sorted_left), len(merged), tmp)

    return merged
 
# Main
if __name__ == "__main__":

    if len(sys.argv) != 3:
        print("Usage: python mergesort.py SIZE THRESHOLD")
        MPI.Finalize()
        sys.exit(1)

    N = int(sys.argv[1])
    threshold = int(sys.argv[2])
    max_depth = math.floor(math.log2(size))  # the max depth to split to ranks

    if rank == 0:
        start_gen = time.time()  
        # Generate reverse sorted array  
        arr = [N - i for i in range(N)]

        start_sort = time.time()
        sorted_arr = merge_sort_parallel(arr, threshold, max_depth, 0)
        end = time.time()

        # Correctness check: H = Sum^{N-1}_{i=0} (i * arr[i}) <=> H = N(N-1)(N+1)/3
        H = 0
        for i, val in enumerate(sorted_arr):
            H += i * val
        c = H / (N * N * N)   # scaled correctness check

        duration_ete = end - start_gen
        duration_sort = end - start_sort

        print(f'{duration_ete},{duration_sort},{c:.16f}')

        # Tell all workers to stop
        for r in range(1, size):
            comm.send([], dest=r, tag=999)

    else:
        status = MPI.Status()
        # wait for message from parent
        while True:
            data = comm.recv(source=MPI.ANY_SOURCE, tag=MPI.ANY_TAG, status=status)
            parent = status.Get_source()
            tag = status.Get_tag()

            if tag == 999:
                break

            depth = tag
            sorted_data = merge_sort_parallel(data, threshold, max_depth, depth + 1)
            comm.send(sorted_data, dest=parent, tag=depth)

    MPI.Finalize()

